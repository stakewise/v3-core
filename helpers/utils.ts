import { Contract } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/src/types/runtime'

export async function deployContract(tx: any): Promise<Contract> {
  const result = await tx
  await result.deployTransaction.wait()
  return result
}

async function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function verify(
  hre: HardhatRuntimeEnvironment,
  address: string,
  constructorArgs: any,
  contractPath: string
) {
  if (!process.env.BLOCK_EXPLORER_KEY) {
    return
  }

  let count = 0
  const maxTries = 8
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await delay(10000)
    try {
      console.log('Verifying contract at', address)
      await hre.run('verify:verify', {
        address,
        constructorArguments: constructorArgs,
        contract: contractPath,
      })
      break
    } catch (error) {
      if (String(error).includes('Already Verified')) {
        console.log(`Already verified contract at ${contractPath} at address ${address}`)
        break
      }
      if (++count == maxTries) {
        console.log(
          `Failed to verify contract at ${contractPath} at address ${address}, error: ${error}`
        )
        break
      }
      console.log(`Retrying... Retry #${count}, last error: ${error}`)
    }
  }
}
