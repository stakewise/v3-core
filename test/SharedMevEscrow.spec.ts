import { ethers, waffle } from 'hardhat'
import { Wallet } from 'ethers'
import { SharedMevEscrow, VaultsRegistry } from '../typechain-types'
import snapshotGasCost from './shared/snapshotGasCost'
import { expect } from './shared/expect'
import { parseEther } from 'ethers/lib/utils'
import { ethVaultFixture } from './shared/fixtures'

const createFixtureLoader = waffle.createFixtureLoader

describe('SharedMevEscrow', () => {
  let owner: Wallet, other: Wallet, sharedMevEscrow: SharedMevEscrow, vaultsRegistry: VaultsRegistry

  let loadFixture: ReturnType<typeof createFixtureLoader>

  before('create fixture loader', async () => {
    ;[owner, other] = await (ethers as any).getSigners()
    loadFixture = createFixtureLoader([owner])
  })

  beforeEach('deploy fixture', async () => {
    ;({ sharedMevEscrow, vaultsRegistry } = await loadFixture(ethVaultFixture))
  })

  it('vault deployment gas', async () => {
    const sharedMevEscrowFactory = await ethers.getContractFactory('SharedMevEscrow')
    const tx = await sharedMevEscrowFactory.deploy(vaultsRegistry.address)
    await snapshotGasCost(tx.deployTransaction)
  })

  it('only vault can withdraw assets', async () => {
    await expect(sharedMevEscrow.connect(other).harvest(parseEther('1'))).to.be.revertedWith(
      'HarvestFailed'
    )
  })

  it('emits event on transfers', async () => {
    const value = parseEther('1')

    await expect(
      other.sendTransaction({
        to: sharedMevEscrow.address,
        value: value,
      })
    )
      .to.be.emit(sharedMevEscrow, 'MevReceived')
      .withArgs(value)
  })
})
