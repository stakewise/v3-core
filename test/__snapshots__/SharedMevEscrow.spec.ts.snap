// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedMevEscrow vault deployment gas 1`] = `
Object {
  "calldataByteLength": 700,
  "gasUsed": 170437,
}
`;
